src/Juvix/Frontend/FreeVars.hs

```diff
@@ -0,0 +1,108 @@
+import qualified Juvix.Core.Common.NameSymbol as NameSymbol
+import Data.HashSet (HashSet, difference)
+import GHC.Generics
+
+(\\) :: (Eq a, Hashable a) => HashSet a -> HashSet a -> HashSet a
+(\\) = difference
+
+
+type VarSet = HashSet NameSymbol.T
+
+
+class FreeVars a where
+  freeVars :: a -> VarSet
+  default freeVars :: (Generic a, GFreeVars (Rep a)) => a -> VarSet
+  freeVars = gfreeVars . from
+
+
+type ExpressionAll' c ext =
+  (ExpressionAll c ext,
+   CondAll c ext (Expression' ext),
+   LetAll c ext,
+   ModuleEAll c ext,
+   LetTypeAll c ext,
+   MatchAll c ext,
+   ModuleOpenExprAll c ext,
+   LambdaAll c ext,
+   ApplicationAll c ext,
+   ListAll c ext,
+   TupleAll c ext,
+   BlockAll c ext,
+   InfixAll c ext,
+   ExpRecordAll c ext,
+   DoAll c ext,
+   ArrowExpAll c ext,
+   NamedTypeAll c ext,
+   TypeRefineAll c ext,
+   UniverseExpressionAll c ext,
+   TypeAll c ext)
+
+instance ExpressionAll' FreeVars ext => FreeVars (Expression' ext)
+
+instance (ExpressionAll' FreeVars ext,
+          CondAll FreeVars ext a,
+          FreeVars a)
+  => FreeVars (Cond' ext a)
+
+instance FreeVars (Constant' ext) where freeVars _ = mempty
+
+instance ExpressionAll' FreeVars ext => FreeVars (ModuleE' ext) where
+  freeVars (ModE' binds body ext) =
+    freeVars (binds, body, ext) \\ boundVars binds
+
+instance (ExpressionAll' FreeVars ext, FreeVars a)
+  => FreeVars (CondLogic' ext a)
+
+instance ExpressionAll' FreeVars ext => FreeVars (Let' ext) where
+  freeVars (Let''' binds body ext) =
+    freeVars (binds, body, ext) \\ boundVars binds
+
+instance ExpressionAll' FreeVars ext => FreeVars (LetType' ext) where
+  freeVars (LetType''' binds body ext) =
+    freeVars (binds, body, ext) \\ boundVars binds
+
+-- TODO: other syntax instances of FreeVars
+-- * they can be empty unless there is some binding structure
+-- * ones with a parameter need a more complex context
+--   (like for @Cond'@ above)
+-- * ExpressionAll' will probably also need some more XyzAlls added
+
+
+instance FreeVars a => FreeVars [a]
+instance FreeVars a => FreeVars (NonEmpty a)
+instance FreeVars a => FreeVars (Maybe a)
+instance (FreeVars a, FreeVars b) => FreeVars (Either a b)
+instance (FreeVars a, FreeVars b) => FreeVars (a, b)
+instance (FreeVars a, FreeVars b, FreeVars c) => FreeVars (a, b, c)
+
+
+class GFreeVars f where gfreeVars :: f t -> VarSet
+
+instance GFreeVars V1 where gfreeVars x = case x of {}
+instance GFreeVars U1 where gfreeVars _ = mempty
+
+instance (GFreeVars f, GFreeVars g) => GFreeVars (f :+: g) where
+  gfreeVars (L1 x) = gfreeVars x
+  gfreeVars (R1 y) = gfreeVars y
+
+instance (GFreeVars f, GFreeVars g) => GFreeVars (f :*: g) where
+  gfreeVars (x :*: y) = gfreeVars x <> gfreeVars y
+
+instance GFreeVars f => GFreeVars (M1 i t f) where
+  gfreeVars (M1 x) = gfreeVars x
+
+instance FreeVars c => GFreeVars (K1 i c) where
+  gfreeVars (K1 x) = freeVars x
+
+
+class BoundVars a where boundVars :: a -> VarSet
```

Comment:
I'm not fully sure how all these typeclasseses interact just by reading it, is . It seems we need to do this so we don't fix this pass on a specific version of the frontend?

Reply:
Basically:

- `FreeVars` instances are needed for everything that can occur in an expression, i.e. until GHC stops saying `Could not deduce (FreeVars (Whatever' ext))`
- `FreeVars` can be empty unless there is some binding going on; the default implemented by `GFreeVars` is just to recurse into subterms and concatenate them together.
- `ExpressionAll'` will need to have a constraint added for each datatype reachable from `Expression'`. Sorry about that :( 
- an instance of `BoundVars` will be needed for everything in binding position, returning the names being bound here and visible to the outside (e.g. the definition name in a `Let`, the type name in `LetType`, but *not* the functions'/types' arguments).